### DISCUSSION TITLE
Company-mode based auto-completion for QE-modes.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [company-mode/company-mode#1247] [General]
[QE-modes](http://pwtk.ijs.si/qe-modes.html) is a package which supplies the Emacs major-modes for [Quantum ESPRESSO](http://www.quantum-espresso.org/). The auto-completion mechanism currently implemented in QE-modes is based on the interactive command triggered by `M-x' as noted [here](http://pwtk.ijs.si/qe-modes_user_guide.pdf). Considering that the most popular Emacs completion framework is company-mode, which aims to be a modular in-buffer completion. It will pop up a dynamically generated drop-down box containing the appropriate candidates, generated as you type. So I think it would be great if qe-modes can support auto-completion based on company-mode.

Any tips on implementing this feature for qe-modes package would be appreciated.

Regards,
HZ

**Comment by dgutov** [comment]
I don't really understand how QE-modes work, from the manual.

If you want to have in-buffer completion with Company when using one of those modes, someone should write a backend (or completion function). That sounds like a feature request for the author of QE-modes.

If you want Company to be used for `M-x` completion (in the minibuffer), it's on my list, though pretty far down.

**Reply by hongyi-zhao** [reply]
> someone should write a backend (or completion function).

Got the idea.

> If you want Company to be used for `M-x` completion (in the minibuffer),

Do you mean that this feature is still unavailable in `company-mode` for the moment?

**Reply by dgutov** [reply]
Nope.

It's not "in-buffer completion", after all.

**Reply by hongyi-zhao** [reply]
I still don't quite understand what you mean. To be more specific, I think the completion manner implemented by `company-mode` is more efficient than the `M-x` completion. So I want to implement the in-buffer company completion support for QE-modes, which is the main reason why I ask this question.

**Reply by yugaego** [reply]
> I still don't quite understand what you mean.

Hi @hongyi-zhao , see this [EmacsWiki](https://www.emacswiki.org/emacs/Completion) article to better understand *text* vs *input* completions.  The article isn't 100% purely precise and there're plans to support Company in the minibuffer (I believe Dmitry mentioned [this wishlist item](https://github.com/company-mode/company-mode/issues/144)), but it supposedly might provide you a better understanding.

> So I want to implement the in-buffer company completion support for QE-modes, which is the main reason why I ask this question.

Take a look at the contents of the CONTRIBUTING.md for the starting hints and referenced examples: [Backends Integration](https://github.com/company-mode/company-mode/blob/master/CONTRIBUTING.md#backend-integration) and [Guidelines for Third-Party Packages](https://github.com/company-mode/company-mode/blob/master/CONTRIBUTING.md#guidelines-for-third-party-packages).

**Reply by hongyi-zhao** [reply]
> Take a look at the contents of the CONTRIBUTING.md for the starting hints and referenced examples: [Backends Integration](https://github.com/company-mode/company-mode/blob/master/CONTRIBUTING.md#backend-integration) and [Guidelines for Third-Party Packages](https://github.com/company-mode/company-mode/blob/master/CONTRIBUTING.md#guidelines-for-third-party-packages).

Thank you for telling me these wonderful guides. I want to add the company-mode official [website](http://company-mode.github.io/) and [this wonderful blog](http://sixty-north.com/blog/series/how-to-write-company-mode-backends.html) written by [Austin Bingham](https://github.com/abingham) as a supplement. 

The ultimate purpose of [`QE-modes`](http://pwtk.ijs.si/qe-modes.html) package is to facilitate the generation of [`Quantum ESPRESSO`](http://www.quantum-espresso.org/) input files. Therefore, the content to be automatically completed here may not be the same as ordinary automatic completion. Let me use [the following example](https://github.com/QEF/q-e/blob/master/test-suite/cp_h2o/h2o-mt-blyp-1.in) to illustrate some of the technical difficulties that need to be resolved.


```
&control
   title = ' Water Molecule ',
   calculation = 'cp',
   restart_mode = 'from_scratch',
   ndr = 51,
   ndw = 51,
   nstep  = 100,
   iprint = 100,
   isave  = 100,
   tstress = .TRUE.,
   tprnfor = .TRUE.,
   dt    = 5.0d0,
   etot_conv_thr = 1.d-9,
   ekin_conv_thr = 1.d-4,
   prefix = 'h2o'
   verbosity = 'medium'
/
&system
   ibrav = 14,
   ce
   celldm(1) = 12.0,
   celldm(2) = 1.0,
   celldm(3) = 1.0,
   celldm(4) = 0.0,
   celldm(5) = 0.0,
   celldm(6) = 0.0,
   nat  = 3,
   ntyp = 2,
   nbnd = 4,
   ecutwfc = 80.0,
/
&electrons
   emass = 400.d0,
   emass_cutoff = 2.5d0,
   electron_dynamics = 'damp',
   electron_damping = 0.2
/
&ions
   orthogonalization = 'ortho',
   ion_dynamics = 'none',
   ion_radius(1) = 0.8d0,
   ion_radius(2) = 0.8d0,
/
ATOMIC_SPECIES
   O 16.0d0 O.blyp-mt.UPF
   H 1.00d0 H.blyp-vbc.UPF
ATOMIC_POSITIONS (bohr)
   O     0.0099    0.0099    0.0000  0 0 0
   H     1.8325   -0.2243   -0.0001  1 1 1
   H    -0.2243    1.8325    0.0002  1 1 1
```
The input file specification of the above example is located [here](https://www.quantum-espresso.org/Doc/INPUT_CP.html). The main problems I want to solve are:

1. When I try to input one namelist, hinting all possible namelists at the point with drop-down frame; once I select the namelist, say,  `&CONTROL`, then insert it corresponding to the specification as follows, and put the point in the middle of it, denoted by `|`:

```
&control
|
/
```

2. When I try to input one variable with the point located in certain namelist, hinting all possible variables belonging to the given namelist at the point with drop-down frame. Say, for the namelist `&control`, only [the following variables](https://www.quantum-espresso.org/Doc/INPUT_CP.html#idm4) should be used to construct the candidates list:

`     calculation | title | verbosity | isave | restart_mode | nstep | iprint | tstress | tprnfor | dt | outdir | saverho | prefix | ndr | ndw | tabps | max_seconds | etot_conv_thr | forc_conv_thr | ekin_conv_thr | disk_io | memory | pseudo_dir | tefield `

3. When I try to input one CARD, hinting all possible CARDs at the point with drop-down frame; once I select the CARD, say,  `ATOMIC_SPECIES`, then insert it with an example usage of the CARD, say, the following one:

```
ATOMIC_POSITIONS (bohr)
   O     0.0099    0.0099    0.0000  0 0 0
```
If no further usage code snippets are available, put the point at the beginning of the next line, denoted by `|`, as follows: 

``` 
ATOMIC_SPECIES
|
```

**Reply by dgutov** [reply]
Doesn't sound too alien for in-buffer completion: you look around to see in which part of the file you are, and suggest appropriate completions. Implement `post-completion` action as well, to insert the appropriate additional text together with the completion and move point where you want it to be.

Perhaps creating such files with `M-x` will remain preferable for you. But it doesn't look incompatible with in-buffer completion.

**Reply by hongyi-zhao** [reply]
> you look around to see in which part of the file you are, and suggest appropriate completions. 

Are there any built-in company features for this purpose?

> Implement `post-completion` action as well, to insert the appropriate additional text together with the completion and move point where you want it to be.

Do you mean something similar to `annotation` and `meta` as represented [here](https://github.com/company-mode/company-mode/wiki/Writing-backends#advanced-example)?

> Perhaps creating such files with `M-x` will remain preferable for you. But it doesn't look incompatible with in-buffer completion.

How can  I implement both in one code simultaneously?

**Reply by dgutov** [reply]
> Are there any built-in company features for this purpose?

No, that's simply a task people use Elisp for. Try `back-to-indentation`, `looking-at`, `re-search-backward` and related functions.

> Do you mean something similar to annotation and meta as represented here?

Yes, but `post-completion`. See the `company-backends` docstring for more explanation and the built-in backends for some examples.

> How can I implement both in one code simultaneously?

Not sure what kind of answer you are looking for, but maybe try using `company-begin-with`.
